test/unit/org/apache/cassandra/utils/memory/MemtableCleanerThreadTest.java

```diff
@@ -0,0 +1,198 @@
+    }
+
+    private void stopThread()
+    {
+        cleanerThread.shutdownNow();
+
+        for (int i = 0; i < TIMEOUT_MILLIS && cleanerThread.isAlive(); i++)
+            FBUtilities.sleepQuietly(1);
+
+        assertFalse(cleanerThread.isAlive());
+    }
+
+    private void waitForPendingTasks()
+    {
+        // wait for a bit because the cleaner latch completes before the pending tasks are decremented
+        FBUtilities.sleepQuietly(1);
+
+        for (int i = 0; i < TIMEOUT_MILLIS; i++)
+            FBUtilities.sleepQuietly(1);
+
+        assertEquals(Long.valueOf(0), metric.numPendingTasks.getValue());
+    }
+
+    @Test
+    public void testCleanerInvoked() throws Exception
+    {
+        CountDownLatch cleanerExecutedLatch = new CountDownLatch(1);
+        CompletableFuture<Boolean> fut = new CompletableFuture<>();
+        AtomicBoolean needsCleaning = new AtomicBoolean(false);
+
+        when(pool.needsCleaning()).thenAnswer(invocation -> needsCleaning.get());
+
+        when(cleaner.clean()).thenAnswer(invocation -> {
+            needsCleaning.set(false);
+            cleanerExecutedLatch.countDown();
+            return fut;
+        });
+
+        // start the thread with needsCleaning returning false, the cleaner should not be invoked
+        needsCleaning.set(false);
+        startThread();
+        cleanerExecutedLatch.await(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        assertEquals(1, cleanerExecutedLatch.getCount());
+        assertEquals(Long.valueOf(0), metric.numPendingTasks.getValue());
+
+        // now invoke the cleaner
+        needsCleaning.set(true);
+        cleanerThread.trigger();
+        cleanerExecutedLatch.await(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        assertEquals(0, cleanerExecutedLatch.getCount());
+        assertEquals(Long.valueOf(1), metric.numPendingTasks.getValue());
+
+        // now complete the cleaning task
+        needsCleaning.set(false);
+        fut.complete(true);
+        waitForPendingTasks();
+
+        stopThread();
+    }
+
+    @Test
+    public void testCleanerError() throws Exception
+    {
+        AtomicReference<CountDownLatch> cleanerLatch = new AtomicReference<>(new CountDownLatch(1));
+        AtomicReference<CompletableFuture<Boolean>> fut = new AtomicReference<>(new CompletableFuture<>());
+        AtomicBoolean needsCleaning = new AtomicBoolean(false);
+        AtomicInteger numTimeCleanerInvoked = new AtomicInteger(0);
+
+        when(pool.needsCleaning()).thenAnswer(invocation -> needsCleaning.get());
+
+        when(cleaner.clean()).thenAnswer(invocation -> {
+            needsCleaning.set(false);
+            numTimeCleanerInvoked.incrementAndGet();
+            cleanerLatch.get().countDown();
+            return fut.get();
+        });
+
+        // start the thread with needsCleaning returning true, the cleaner should be invoked
+        needsCleaning.set(true);
+        startThread();
+        cleanerLatch.get().await(TIMEOUT_SECONDS, TimeUnit.SECONDS);
```

Comment:
nit: Should we `assertTrue()` on all these `await()` calls to make sure we fail right away on a timeout?

Reply:
Okay, it looks like some of them should be `assertFalse()` when we're expecting the cleaner not to be invoked ;)